From bb3bcb2498335637ef98f36ab2633026bf8551c7 Mon Sep 17 00:00:00 2001
From: Sai Surya Teja <sssteja@habana.ai>
Date: Thu, 23 Sep 2021 10:11:29 +0300
Subject: [PATCH] [SW-35427] preboot: size check for preboot binary

This patch adds checks for preboot binary size in greco-arc,
gaudi2-arc & gaudi2-arm.
Preboot build fails for binary size > (64-1)KB.
1KB added as a guard to have time to take action if overflows.

This is needed to ensure that preboot binary size doesn't overflow
beyond preboot boundaries in spi flash layout and throws error to user
at build time.

Change-Id: I1a2b7c50a1bc85d734bdba52291db5dfa8b8029b
---

diff --git a/Makefile b/Makefile
index 2c1c9c4..b783942 100644
--- a/Makefile
+++ b/Makefile
@@ -93,6 +93,13 @@
 		exit 1; \
 	fi
 
+image_validation_preboot: preboot
+	@sz=`stat -Lc "%s" $(PREBOOT_TARGET)`; \
+	if [ $$sz -gt $(PREBOOT_MAX_SIZE) ]; then \
+		echo "$(PREBOOT_TARGET) image size ($$sz) is too big (> $(PREBOOT_MAX_SIZE))"; \
+		exit 1; \
+	fi
+
 $(BINDIR)/%.bin: %
 	$(OBJCOPY) -O binary -R .note.gnu.build-id $(BINDIR)/$< $@
 
diff --git a/arch/gaudi2/arc/Makefile b/arch/gaudi2/arc/Makefile
index c2c8bcc..74d9197 100644
--- a/arch/gaudi2/arc/Makefile
+++ b/arch/gaudi2/arc/Makefile
@@ -2,8 +2,11 @@
 
 FW_SIZE := 0x8000
 PPBOOT_MAX_SIZE := 16384
+PREBOOT_MAX_SIZE := 64512
 
 LDFLAGS += -Wl,--just-symbols $(PREBOOT_ROOT_DIR)/../arcboot-custom/exports.elf
 
 PREBOOT_SRCS_C += $(shell ls $(PREBOOT_ROOT_DIR)/lib/preboot/thermal_v2.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/security.c)
+
+image_validation: image_validation_preboot
diff --git a/arch/gaudi2/arm/Makefile b/arch/gaudi2/arm/Makefile
index 838a8c4..e2d6aa8 100644
--- a/arch/gaudi2/arm/Makefile
+++ b/arch/gaudi2/arm/Makefile
@@ -2,8 +2,10 @@
 
 FW_SIZE := 0x4400
 PPBOOT_MAX_SIZE := 20480
+PREBOOT_MAX_SIZE := 64512
 
 PREBOOT_SRCS_C += $(shell ls $(PREBOOT_ROOT_DIR)/lib/preboot/thermal_v2.c)
 
 CFLAGS += -mstrict-align
 image_validation: image_validation_common
+image_validation: image_validation_preboot
diff --git a/arch/greco/arc/Makefile b/arch/greco/arc/Makefile
index 3f8c5b6..33d7b4c 100644
--- a/arch/greco/arc/Makefile
+++ b/arch/greco/arc/Makefile
@@ -1,7 +1,10 @@
 include $(HL_DEV_ARCH_COMMON)/Makefile
 
 FW_SIZE := 0x8000
+PREBOOT_MAX_SIZE := 64512
 
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/pll.c)
 PREBOOT_SRCS_C += $(shell ls $(PREBOOT_ROOT_DIR)/lib/preboot/thermal_v2.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/security.c)
+
+image_validation: image_validation_preboot
